packages/pnpm-sync-lib/src/utilities.ts

```diff
@@ -6,3 +10,40 @@
 export function pnpmSyncGetJsonVersion(): string {
   return require('../package.json').version;
 }
+
+export function getFilesInDirectory(directory: string, includeDirectory: boolean): Array<IFileStat> {
+  const returnFileList: Array<IFileStat> = [];
+  getFilesInDirectoryHelper(directory, includeDirectory, returnFileList);
+  return returnFileList;
+}
+
+function getFilesInDirectoryHelper(
+  directory: string,
+  includeDirectory: boolean,
+  returnFileList: Array<IFileStat>
+): void {
+  const fileList = fs.readdirSync(directory);
+
+  for (const fileName of fileList) {
+    const absoluteFileName: string = path.join(directory, fileName);
+    const fileStat = fs.statSync(absoluteFileName);
+    if (fileStat.isDirectory()) {
```

Comment:
Using the `{ withFileTypes: true }` option to `readdirSync` will save you a bunch of stat calls for all the directories. Also you don't need to stat the files until such time as it is determined that they already exist in the destination folder.

Reply:
My advice on how to do the mirroring would be to write a function that takes a source folder and destination folder and makes the destination folder match the source folder as follows:

```ts
function mirrorFolder(sourceFolder, destinationFolder): void {
// TODO: Wrap the fs.readdirSync calls in a handler for ENOENT/ENOTDIR:
// For destination, ENOTDIR means delete folder (ignoring failure), then:
// For destination, ENOENT means create folder, then return as content []

const sourceItems = fs.readdirSync(sourceFolder, { withFileTypes: true });
const destinationItems = fs.readdirSync(destinationFolder, { withFileTypes: true });

const sourceItemsByName = new Map(sourceItems.map(entry => [entry.name, entry]));
const destinationItemsByName = new Map(destinationItems.map(entry => [entry.name, entry]));

for (const [name, sourceItem] of sourceItemsByName) {
  const sourcePath = path.join(sourceFolder, name);
  const destinationPath = path.join(destinationFolder, name);

  const destinationItem = destinationItemsByName.get(name);
  if (destinationItem) {
    destinationItemsByName.delete(name);
    if (sourceItem.isFile() && destinationItem.isFile()) {
      // only case in which we need to stat() anything
      const sourceStat = fs.statSync(sourcePath, { throwIfNoEntry: false });
      const destinationStat = fs.statSync(destinationPath, { throwIfNoEntry: false });
      
      if (sourceStat.ino === destinationStat.ino) {
        continue;
      }
    } else if (sourceItem.isDirectory() && destinationItem.isDirectory()) {
      mirrorFolder(sourcePath, destinationPath);
      continue;
    } else if (sourceItem.isSymbolicLink()) {
      throw new Error(`Source folder contains a symbolic link: "${sourcePath}". This is not supported for pnpm-sync because it leaks data across the injection boundary.`);
    }

    // Existing item is wrong type, or has wrong inode#
    fs.rmSync(destinationPath, { force: true, recursive: destinationItem.isDirectory() });

    if (sourceItem.isFile()) {
      fs.linkSync(sourcePath, destinationPath);
    } else {
      fs.mkdirSync(destinationPath, { recursive: true });
      mirrorFolder(sourcePath, destinationPath);
    }
  }
}

// This collection now only contains entries that don't exist in the source
for (const [name, destinationItem] of destinationItemsByName) {
  const destinationPath = path.join(destinationFolder, name);
  fs.rmSync(destinationPath, { force: true, recursive: destinationItem.isDirectory() });
}
}
```

For performance, this should be run in a worker thread. On Linux the performance of using synchronous calls in a worker is faster than async calls. On Windows it should average out. Darwin may be an outlier.
Note that this implementation assumes that no symbolic links will be mirrored, since symbolic links would cause data leakage from the source folder.

Reply:
Thanks for all suggestions. I refactored the code a little bit based on your `mirrorFolder` logic.

Regarding the synchronous vs asynchronous for file operations, I did an experiment and found out the synchronous operations is slower than asynchronous. 

An example, a folder that has ~3500 files.
The asynchronous copy from scratch will cost around ~700ms, while the synchronous copy from scratch will cost around ~1700ms

So, I am thinking, since we already named this copy function as `pnpmSyncCopyAsync`, we will keep it constant, use async file operations in this function. 

Later, if we think a synchronous version is needed (for example, to be used by a Node.js worker), we can provide a separated synchronous version of `pnpmSyncCopy`.